src/garage/_dtypes.py

```diff
@@ -625,6 +625,27 @@ def last(self):
         return self.step_type is StepType.TERMINAL or self.step_type \
             is StepType.TIMEOUT
 
+    @classmethod
```

Comment:
Doing all the validation above on every step will get really expensive. I think we should leave validation for the batch types (`TimeStepBatch` and `TrajectoryBatch`).

Please remove the validation parts of `__new__`